styleguide/drafts/tooling.rst

```diff
@@ -0,0 +1,518 @@
+Admittedly it's powerfulness comes with a cost of complexity for bigger setups.
+Still the overall the benefits outweigh the cost, and with the detail in mind that
+lots of projects are using sphinx it is worth the effort to learn and use it.
+
+Pros & Cons
+~~~~~~~~~~~
+
+.. list-table::
+    :header-rows: 1
+    :widths: 50 50
+
+    * - Pros ✅
+      - Cons ❌
+    * - Widely used
+      - Additional dependency
+    * - Lots of extensions
+      - Restructured text can be quirky at times
+    * - Lots of output formats are supported
+      - Lots to learn/know regarding .rst and sphinx
+    * - Basics are easy to learn
+      -
+    * - Extensions for api documentation
+      -
+    * - Plugin mechanism
+      -
+
+Alternatives
+~~~~~~~~~~~~
+
+* `mkdocs <https://github.com/mkdocs/mkdocs>`_
+* `pydoc <https://pdoc.dev/docs/pdoc.html>`_
+
+
+Dependencies & Packaging
+________________________
+
+.. figure:: https://imgs.xkcd.com/comics/dependency.png
+   :alt: xkcd-2347
+   :target: https://xkcd.com/2347/
+
+   source: `xkcd.com <xkcd_>`_
+   license: `CC BY-NC 2.5`_
+
+
+Poetry_
++++++++
+`Project Homepage <Poetry_www_>`_
+
+Poetry has a very good package/version resolver and simplifies packaging and updating dependencies significantly.
+
+Pros & Cons
+~~~~~~~~~~~
+
+.. list-table::
+    :header-rows: 1
+    :widths: 50 50
+
+    * - Pros ✅
+      - Cons ❌
+    * - Good dependency resolver
+      - Fairly new, some edge cases my not supported yet
+    * - Uses pyproject.toml for configuration
+      - Toml file definition is less flexible than setup.py based one
+    * - Takes care of project versioning
+      - Editable and/or repo based install(s) are not possible "out of the box"?
+    * - Dependency pinning
+      - setup.py is still required for installations with older versions of pip
+    * - Very active development & community
+      -
+    * - Good & Powerful CLI
+      -
+    * - No manual dependency editing required
+      -
+
+Alternatives
+~~~~~~~~~~~~
+
+* `pipenv <https://pipenv.pypa.io/en/latest/>`_
+
+
+Code Health
+___________
+
+.. figure:: https://imgs.xkcd.com/comics/code_quality.png
+   :alt: xkcd-2347
+   :target: https://xkcd.com/2347/
+
+   source: `xkcd.com <xkcd_>`_
+   license: `CC BY-NC 2.5`_
+
+Pylint_
++++++++
+`Project Homepage <Pylint_www_>`_
+
+Helps finding bugs and issues before they are a problem and improves overall code quality.
+
+Pros & Cons
+~~~~~~~~~~~
+
+.. list-table::
+    :header-rows: 1
+    :widths: 50 50
+
+    * - Pros ✅
+      - Cons ❌
+    * - Lots of good warnings and hints
+      - Extra dependency
+    * - Extension mechanism (custom extensions e.g. perflint)
+      - Extra learning curve not all messages maybe straight forward
+    * - Rating check 0-10 makes it possible to steadily improve a code base
+      -
+    * - Compatible with sonar
+      -
+    * - Supports pyproject.toml
+      -
+
+Alternatives
+~~~~~~~~~~~~
+
+* `Flake8 <https://flake8.pycqa.org/en/latest/>`_
```

Comment:
what is the benefit of pylint to the alternatives

Reply:
In the end most of them are quite the same and they have overlaping features.
I personally have the most experience with `pylint` and I also like the following features of it:
* "rating" feature which helps to improve the code base over time without imposing super strict restrictions. 
   (one can commit as long as overall code quality does not decrease, without disallowing "new warnings".
   e.g. if you fix 2 warnings and introduce a new one, you are good to commit it)
* Lots helpful warnings (e.g. dangerous default parameters `def func(a=[])`)
* Plugin/Extension point ("if custom checks etc. are needed")

Reply:
Sounds good, I used also pylint so far, but didn't look too much into the other ones and didn't have the comparison

Reply:
so bullet point 4. flat learning curve for at least 2 python devs ;)

Reply:
Note: 
I also never did a in detail comparison, but from my experience pylint does its job well.